adapter/service/ota/service.go

```diff
@@ -0,0 +1,161 @@
+package ota
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/futurehomeno/fimpgo"
+	"github.com/futurehomeno/fimpgo/fimptype"
+
+	"github.com/futurehomeno/cliffhanger/adapter"
+)
+
+// Controller is an interface representing a device capable of OTA firmware updated.
+// In a polling scenario implementation might require some safeguards against excessive polling.
+type Controller interface {
+	// StartOTAUpdate starts an OTA update process with provided firmware path.
+	StartOTAUpdate(firmwarePath string) error
+
+	// OTAUpdateReport returns an OTA update report.
+	OTAUpdateReport() (UpdateReport, error)
+}
+
+// Service is an interface representing an OTA FIMP service.
+type Service interface {
+	adapter.Service
+
+	// StartUpdate starts an OTA update process with provided firmware path.
+	StartUpdate(firmwarePath string) error
+
+	// SendStatusReport sends an OTA status report.
+	// If controller reports idle status, no report is sent.
+	SendStatusReport() error
+}
+
+// Config represents a service configuration.
+type Config struct {
```

Comment:
I think the service implementation needs to be different. IMHO It needs methods (besides StartUpdate):

SendStartReport
SendProgressReport
SendEndReport

Entire logic of triggering those methods must be on the controller implementation.

There are no getters for these.